Use artifact MUMPS_jll

amontoison opened the pull request:
close #53

dpo commented:
Let's not remove the old interface until everything works. That will also help keep this PR more focused.

amontoison commented:
> Let's not remove the old interface until everything works. That will also help keep this PR more focused.

I update the PR to not remove the old interface except `deps/build.jl`.

I added a `ci2.yml` to test MUMPS.jl with MUMPS_jll because I updated the interface such that the users can use their own installation of MUMPS if the environment variable `MUMPS_PREFIX` is defined.
All tests with `ci_yml` still use your tap dpo/mumps-jl.

dpo commented:
> I added a ci2.yml to test MUMPS.jl with artifact MUMPS_jll because I updated the interface such that the users can use their own installation of MUMPS if the environment variable MUMPS_PREFIX is defined.
All tests with ci.yml still use your tap dpo/mumps-jl.

Right, but it can no longer be built...

Also, why do we need to install all three oh OpenMPI, MPICH and MicrosoftMPI???

amontoison commented:
> > I added a ci2.yml to test MUMPS.jl with artifact MUMPS_jll because I updated the interface such that the users can use their own installation of MUMPS if the environment variable MUMPS_PREFIX is defined.
> > All tests with ci.yml still use your tap dpo/mumps-jl.
> 
> Right, but it can no longer be built...

I found a solution to keep it.
 
> Also, why do we need to install all three oh OpenMPI, MPICH and MicrosoftMPI???

OpenMPI, MPICH and MicrosoftMPI are dependencies of MPI.jl (https://github.com/JuliaParallel/MPI.jl/blob/master/Project.toml#L10-L13).
They are required for MPItrampoline_jll.

dpo commented:
#81 should fix the test that fails on macOS related to accuracy.

There's still a segfault with the JLL to investigate.

dpo reviewed (commented):
It all looks great, many thanks! I would prefer to leave MUMPS.jl and mumps_types.jl unchanged in this PR and update MUMPS-C_API.jl instead.

dpo commented:
I'll try locally.

dpo commented on src/MUMPS.jl:
```diff
@@ -20,7 +20,7 @@ as `x=mumps_solve(A,y)`, or `mumps_solve!(x,A,y)`.
 work with mumps objects.
 
-Note, unless working with the low-level interace, we discourage setting the `JOB`
+Note, unless working with the low-level interface, we discourage setting the `JOB`
```
```suggestion
Note, unless working with the low-level interace, we discourage setting the `JOB`
```

dpo commented on src/MUMPS.jl:
```diff
@@ -29,6 +29,8 @@ MUMPS has to offer.
 module MUMPS
 
+using MPI, Libdl, LinearAlgebra, SparseArrays
+
```
```suggestion
```

dpo commented on src/MUMPS_C-API.jl:
```diff
@@ -1,24 +1,21 @@
-using Libdl, LinearAlgebra, SparseArrays
-
-using MPI
```
restore this line

dpo commented:
ps: at least, Homebrew is also installing binaries now, which speeds up the CI (https://github.com/JuliaSmoothOptimizers/MUMPS.jl/runs/4811639965?check_suite_focus=true#step:3:59).

pps: MUMPS itself is still being compiled on Linux... not sure why.

amontoison commented:
> ps: at least, Homebrew is also installing binaries now, which speeds up the CI (https://github.com/JuliaSmoothOptimizers/MUMPS.jl/runs/4811639965?check_suite_focus=true#step:3:59).

Great !

> pps: MUMPS itself is still being compiled on Linux... not sure why.

MUMPS is not compiled with my `ci2.yml` (https://github.com/JuliaSmoothOptimizers/MUMPS.jl/runs/4811022469?check_suite_focus=true).
Maybe we should update the build name of `ci.yml` and `ci2.yml`.
Instead of `CI`, we could rename them `CI-Homebrew` and `CI-Artifact`.

dpo commented:
I can reproduce the segmentation fault locally on macOS... :-(

amontoison commented:
@dpo I recompiled MUMPS_jll on my computer for `x86_64-linux-gnu` and `x86_64-apple-darwin` platforms without `ParMETIS` and deployed the new MUMPS_jll (fork amontoison/MUMPS_jll.jl).
We still have the segmentation fault :(

dpo commented:
😞 
It's hard to debug. I tried to follow the execution with `dtrace` and saw messages that suggested ParMETIS as a potential problem. I'll look more when I find a moment.

amontoison commented:
The version 2.2.0 of SCALAPACK was released four days ago: http://www.netlib.org/scalapack/#_scalapack_version_2_2_0
I recompiled it with Yggdrasil: https://github.com/JuliaPackaging/Yggdrasil/pull/4368
It could fix our issue with Mac :crossed_fingers:

amontoison commented:
`SCALAPACK` 2.2.0 doesn't solved our issue but I found a topic about our segfault: https://developer.apple.com/forums/thread/121887
The solution is to recompile with `-fno-stack-check` flag.

dpo commented:
I tried that 10 days ago, but it didn't help. Maybe I didn't recompile Scalapack with it though. I'll have to try.

amontoison commented:
@dpo MUMPS_jll works on Linux and Mac !!! :tada:

dpo commented:
Excellent, thank you! I'm going to merge this despite the failing Homebrew actions. In the next PR, I'll remove the Homebrew actions and add M1.

amontoison commented:
We can also test FreeBSD with CirrusCI :-)
